### ISSUE TITLE
Odd batch_size specific behaviour with nasnet_large on ImageNet validation

### ISSUE THREAD

**Issue by rwightman** [tensorflow/models#2778]
I've been trying to reproduce the ImageNet validation accuracies posted with the pretrained weights.

I cannot reproduce the results with an image size of 331 and the default batch size of 100 used in the eval_image_classifer script. With defaults I get roughly 64% top-1 and 74% top-5 using the command:

`python eval_image_classifier.py --dataset_dir ~/imagenet/ --dataset_split_name validation --model_name nasnet_large --checkpoint_path ~/nasnet_large/model.ckpt  --dataset_name imagenet --moving_average_decay=0.9999`

However, if I change the batch size down to 50, I hit the posted results dead on, 82.7% top-1 and 92.6% top-5. This seems rather odd to me. I've never experienced that significant of a performance drop by changing the batch size by a factor of two at eval. 

I noticed that with the batch size of 100, I could improve performance by dropping the eval image size down. There is an optimal value somewhere in the 280-290 range before it starts to fall off. My best result with batch size of 100 was at 288x288 with a 80.8% top-1 and 95.2% top-5, it falls off quickly as the image size increases from there. 

Is this possibly some sort of overflow in batch norm or another op in the network?

Note, I'm using latest released pip install of Tensorflow 1.4 for GPU. CUDA 8.0, cuDNN 6.

**Comment by chenghuige** [comment]
@rajatmonga , I face this also... Ando when train I use eval_batch_size 50 and eval result during train is as expected. I frozen ops for inference, and when inference from pb model saved and set inference buffer size 100, I got worse inference result. But when set inference batch size to 50 the result is ok.
Is it due to moving average decay 0.9999  as other models like incpetion V4 do not met this problem.
Also even if use batch size 50 for different run I will get slightly different inference result that is confusing.

**Comment by chenghuige** [comment]
1. I verified on two machines, there might be some env idff, running same code inference using same model. I have tested both tf 1.3 and tf 1.4 on one machine batch size 100 and 50 all get cider 1.73, 
on another machine batch size 100 can only get cider 1.49.
2. On both machine, different inference run will get slightly different result.

**Comment by chenghuige** [comment]
Is it the problem related to below code in nasnet.py ? seems is_batchnorm_training is always True, even
when inference, if using nets_factory.py as we only call
 func(images, num_classes, is_training=is_training, **kwargs)

def build_nasnet_large(images, num_classes,
                       is_training=True, is_batchnorm_training=True,
                       final_endpoint=None):
with arg_scope([slim.batch_norm], is_training=is_batchnorm_training):

**Comment by shlens** [comment]
@rwightman @chenghuige 
Does your version of nasnet include this [change](https://github.com/tensorflow/models/commit/cf8f014b82d1126d73c219ba0b5815e8c891ebcc)? This change fixes a bug that left BatchNorm on during evaluation and would lead to lower performance.

**Comment by rwightman** [comment]
@shlens 
Yes, my code includes that change. I synced up with the latest yesterday and also re-downloaded the checkpoints. I expect if that was the issue I would not have full performance with batch 50 and degraded with batch size of 100 but degraded in all cases.

I ran another test overnight, I ran imagenet validation with batch size of 100 on my CPU. It produces the correct result.

**Comment by shlens** [comment]
@rwightman I am surprised by your results. Would you be able to run with batch size = 25 and batch size = 200? Just for clarity, are you running on all 50,000 validation images?

**Comment by rwightman** [comment]
@shlens 
Okay, so I've run some more passes, results not what I was expecting, so had to do more passes including an 'am I going insane' re-run of batch_size=100.  

The problem seems very specific to batch_size=100 on GPU. At a batch_size of 100, performance peaks somewhere around 288-290 image size as opposed to the native 331x331. I've posted results from both yesterday's and now today's runs. All of them on the full 50,000 image ImageNet validation set. 

As an additional check, I ran a validation pass using inception_resnet_v2 pretrained weights @ native 299x299 with a batch_size of 100 using the eval script and it matches expected results.

**Yesterday:**

batch_size=100 @ 256x256
2017-11-12 17:06:31.406271: eval/Accuracy[0.79056]
2017-11-12 17:06:31.406272: eval/Recall_5[0.94308]

batch_size=100 @ 288x288
2017-11-12 18:42:19.251169: eval/Accuracy[0.80766]
2017-11-12 18:42:19.251173: eval/Recall_5[0.95234]

batch_size=100 @ 299x299
2017-11-12 16:46:09.191230: eval/Accuracy[0.77208]
2017-11-12 16:46:09.191243: eval/Recall_5[0.90276]

batch_size=100 @ 331x331
2017-11-12 15:54:28.281820: eval/Accuracy[0.63622]
2017-11-12 15:54:28.281828: eval/Recall_5[0.74138]

batch_size=50 @ 331x331
2017-11-12 19:49:34.019025: eval/Accuracy[0.82706]
2017-11-12 19:49:34.019025: eval/Recall_5[0.96168]

**Today:**

batch_size=25 @ 331x331
2017-11-13 14:43:20.327825: eval/Accuracy[0.82706]
2017-11-13 14:43:20.327826: eval/Recall_5[0.96168]

batch_size=200 @ 331x331
2017-11-13 15:01:58.311747: eval/Accuracy[0.82706]
2017-11-13 15:01:58.311767: eval/Recall_5[0.96168]

batch_size=100 @ 331x331
2017-11-13 15:20:36.148437: eval/Accuracy[0.63712]
2017-11-13 15:20:36.148441: eval/Recall_5[0.74282]

batch_size=99 @ 331x331
2017-11-13 16:19:21.237480: eval/Accuracy[0.82706]
2017-11-13 16:19:21.237482: eval/Recall_5[0.96168]

batch_size=101 @ 331x331
2017-11-13 17:12:32.546138: eval/Accuracy[0.82706]
2017-11-13 17:12:32.546145: eval/Recall_5[0.96168]

**Comment by chenghuige** [comment]
@shlens just updated slim lib and re tested,  still face the two problem I mentioned above.

**Comment by chenghuige** [comment]
@shlens  Also if I'm using old code(with is_batchnorm_traning param) train model, seems I can not change to use new code to finetune from old models, the result will be nearly all random start..

**Comment by rwightman** [comment]
This has truly baffled me. I want to ensure there are no issues before investing more time doing some work with this model so I rsync'd my whole experiment to another machine with same driver/cuda/cudnn versions and ran the eval again. It works, no weird behaviour at batch 100. The GPU isn't exactly the same but it's same family (Pascal) so I doubt that's a factor.

At this point I am thinking I hit one of those lovely driver/state issues. I'm going to reboot the original machine once another task is complete and retry later this afternoon. 

I suspect @chenghuige had issues due to the old batch_norm behaviour due to the random variation in his results. It's likely safe to close this.

**Comment by rwightman** [comment]
Alright, more data. I did manage reproduce on the other machine. I made one very small change before rsyncing, I changed the data format from NHWC to NCHW before I started a training session on the original machine.

The only delta in the diff is this line: https://github.com/tensorflow/models/blob/master/research/slim/nets/nasnet/nasnet.py#L79

If I flip that line back to the original 'NHWC' the batch_size=100 problem exists on both machines. I have just confirmed that it doesn't exist on either machine with that line changed to 'NCHW'.

**Comment by rajatmonga** [comment]
@rwightman thanks for narrowing down the issue.

@vrv Have you looked at the NHWC format support for nasnet?

**Comment by vrv** [comment]
NHWC works for CPU, and NCHW works for GPU.  Based on this evidence, it seems like likely to be a CuDNN bug or a bug in a GPU-specific codepath for convolution in TensorFlow, given it fails at a very specific batch size.

**Comment by rajatmonga** [comment]
@vrv Is it feasible to turn off the untested path/raise error on CPU/GPU so
users don't get incorrect issues?

Also worth creating a separate issue for NHWC on GPU.

**Comment by rwightman** [comment]
On that note, the reason I hit this issue right away is that the default batch size for eval_image_classifier.py is 100, and NHWC is the default data format for both CPU and GPU. 

So, anyone that decides to try Nasnet large out with defaults is likely to hit the problem. Might be worth a quick change to the script while the root problem is figured out.

**Comment by vrv** [comment]
We don't really know how to detect that the GPU is going to be used at graph construction time (at our layer of abstraction).  If we did, we would have already chosen NCHW for GPU and NHWC for CPU.

We can change the default eval batch size to 50 to avoid this problem, if that makes things easier.

**Comment by vrv** [comment]
@rwightman 

One thing you can try doing: set the TF_CUDNN_USE_AUTOTUNE environment variable to 0 when running your program with batch size 100 and see if the problem goes away.  I wonder whether it's caused by a bug in the algorithm chosen by the autotuner.

**Comment by vrv** [comment]
Or more precisely, the autotuner is selecting an algorithm that has a bug for the particular shape we are using.

(One idea that comes to mind is that the autotuner could check for rough output agreement from all algorithms and disqualify the algorithms that differ significantly).  cc @zheng-xq @yzhwang

**Comment by yzhwang** [comment]
Setting the data_format to "NHWC" means we won't do NHWCtoNCHW() and NCHWtoNHWC() before and after each cudnn kernel, that include different types of convolutions as well as batch normalization.
Given that the problem only exists in the above case, another possibility is that these two conversions return incorrect values silently.

@rwightman Is there a minimum reproducer you can provide so that I can try some tests for data format conversion on different shapes to see if there's any bug there.

Trying with environ['TF_CUDNN_USE_AUTOTUNE']='0' is also a good thing to try.

**Comment by zheng-xq** [comment]
We are working extensively with NVIDIA to test the functional correctness for each algorithm on many shapes.

Vijay's proposition would be very useful. I've talked to Yang, and he will be adding that check.

**Comment by rwightman** [comment]
I've set TF_CUDNN_USE_AUTOTUNE=0 in my environment before running and there was no change, still bad results with batch_size=100 and NHWC, good otherwise.

**Comment by zheng-xq** [comment]
@yzhwang, please also look into whether there is a problem with the tensor transpose operations. Thanks

**Comment by yzhwang** [comment]
Thanks @vrv and @zheng-xq , I will reproduce the problem with the command from the original comment.

**Comment by yzhwang** [comment]
OK. Running eval_image_classifier script now with imagenet dataset. I'm not sure how long it will take to get the results, but I will start some debugging once I reproduce the low accuracy results (I know a few places to look at) and give updates once I have some insights.

**Comment by yzhwang** [comment]
Also, I'm wondering if I ever need to wait for the model to converge at all? It would also be good if someone can provide me with instructions on how to get some additional outputs such as these:
2017-11-13 17:12:32.546138: eval/Accuracy[0.82706]
2017-11-13 17:12:32.546145: eval/Recall_5[0.96168]

@rwightman Could you provide more details on how to experiments with different batch_size and image_size quickly? (I know the flags to use, just not sure how you manage to get results in less than one hour for each experiment.

**Comment by vrv** [comment]
This isn't training, this is just running eval on a pretrained model; it shouldn't take too long of a time.

**Comment by rwightman** [comment]
@yzhwang Do you have the pretrained weights for the nasnet large model? There's no convergence, those summaries are dumped at the end of the 500 (for batch size 100) batches.

`python eval_image_classifier.py --dataset_dir /data/imagenet/validation/ --dataset_split_name validation --model_name nasnet_large --checkpoint_path /data/nasnet_large/model.ckpt  --dataset_name imagenet --moving_average_decay=0.9999 --eval_image_size 331 --batch_size 100 `

**Comment by yzhwang** [comment]
Thanks @vrv and @rwightman . I understand how to download the pretrained model and run the script now.

**Comment by yzhwang** [comment]
@rwightman I tried your script with both TensorFlow r1.4 branch built from source (https://github.com/tensorflow/tensorflow/tree/r1.4) and TensorFlow current head built from source and I was able to get the expected results:
2017-11-15 22:55:14.932129: I tensorflow/core/kernels/logging_ops.cc:79] eval/Accuracy[0.82706]
2017-11-15 22:55:14.932129: I tensorflow/core/kernels/logging_ops.cc:79] eval/Recall_5[0.96168]
I'm using a GTX 1080 card (same compute capability 6.1 as Titan Xp and 1080 Ti) with driver version of 387.12.
For tensorflow/models I just used the head of master branch.

I will try r1.4 install version tomorrow.

**Comment by yzhwang** [comment]
OK I tried with the installed version of tensorflow-gpu 1.4. I still got the expected results.
@rwightman Could you report your display driver version? If it is an older one (such as 367) could you try update it to 381, 384, or 387 and try again?

**Comment by rwightman** [comment]
I've got the standard driver for the Ubuntu 16.04 cuda 8 deb which is still
375. I did also spin up a 16.04 p100 in GCP with cuda 9 and a 384 driver
and it also exhibits the same problem. Your driver cuda combo is newer than
any of the default configs.

**Comment by yzhwang** [comment]
I downgraded my driver to 375.82 and ran this:
`python eval_image_classifier.py --dataset_dir ~/imagenet-2012-tfrecord/ --dataset_split_name validation --model_name nasnet_large --checkpoint_path ~/nasnet_large_ckpt/model.ckpt --dataset_name imagenet --moving_average_decay=0.9999 --eval_image_size 331 --batch_size 100`
It still gives me the expected results. (For all these runs, I was using CUDA 8.0 + cudnn v6).

**Comment by rwightman** [comment]
Well, this one is an odd puzzle.

I've reproduced on my initial two local systems with:
1080TI/TitanX - Driver 375.88, CUDA V8.0.61, CUDNN 6.0.21, Python 3.5.1+

I reproduced problem on a GCP instance with:
P100 - Driver 384.90, CUDA V9.0.176, CUDNN 7.0.2, Python 3.5.2

I spun up another cloud instance today. I could not reproduce on an AWS P3 instance with:
V100 - Driver 384.81, CUDA V9.0.176, CUDNN 7.0.3, Python 3.6.3 (Anaconda)

So I've got 3 fails, and one pass on different configs from same code, same dataset, same checkpoints, different hardware, but no clear pattern.

**Comment by yzhwang** [comment]
Hi @rwightman , I tried with Python 3.4 on my 1080 with driver 375 and CUDA 8 + cudnn 6. Still showing expected results (82.7% top one and 96.2% top five).
For CUDA 9 and cudnn 7 version, I assume you built from source? I'm doing this right now on a machine with P100, so that I can match your configuration. I will report my findings later.

**Comment by rwightman** [comment]
I used this wheel for GCP P100 (https://github.com/mind/wheels/releases/tag/tf1.4-gpu-cuda9-generic) .

On AWS, I used their latest AMI which includes a build of TF 1.4. The AMI is advertised as having CUDA 9 and that's what nvcc and libs in the path indicate, however the included module looks like it's linking against CUDA 8.0 based on the ldd output. So it is not in fact CUDA 9.

Is your TF linked against or loading any BLAS libraries? I dumped the output of LDD for two libs in the tensorflow python module and the working V100 variation includes openblas and openmpi whereas the others do not... see next post.

**Comment by rwightman** [comment]
My machine (has issue @ batch 100):

libtensorflow_framework.so
```
linux-vdso.so.1 =>  (0x00007ffda6988000)
libcublas.so.8.0 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcublas.so.8.0 (0x00007f00553cd000)
libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007f00549e4000)
libcudnn.so.6 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.6 (0x00007f004b482000)
libcufft.so.8.0 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcufft.so.8.0 (0x00007f0042633000)
libcurand.so.8.0 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcurand.so.8.0 (0x00007f003e6bd000)
libcudart.so.8.0 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudart.so.8.0 (0x00007f003e457000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f003e253000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f003df4a000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f003dd2d000)
libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f003d9ab000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f003d795000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f003d3cb000)
/lib64/ld-linux-x86-64.so.2 (0x00007f0059262000)
librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f003d1c3000)
libnvidia-fatbinaryloader.so.375.88 => /usr/lib/nvidia-375/libnvidia-fatbinaryloader.so.375.88 (0x00007f003cf77000)

```
contrib/framework/python/ops/_variable_ops.so
```
linux-vdso.so.1 =>  (0x00007ffe9b5de000)
libtensorflow_framework.so => /home/wiggs/work/env/learn3/lib/python3.5/site-packages/tensorflow/contrib/framework/python/ops/../../../../libtensorflow_framework.so (0x00007fb0d665e000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fb0d6355000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fb0d6151000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fb0d5f34000)
librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fb0d5d2c000)
libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fb0d59aa000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb0d55e0000)
libcublas.so.8.0 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcublas.so.8.0 (0x00007fb0d259b000)
libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007fb0d1bb2000)
libcudnn.so.6 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudnn.so.6 (0x00007fb0c8650000)
libcufft.so.8.0 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcufft.so.8.0 (0x00007fb0bf801000)
libcurand.so.8.0 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcurand.so.8.0 (0x00007fb0bb88b000)
libcudart.so.8.0 => /usr/local/cuda-8.0/targets/x86_64-linux/lib/libcudart.so.8.0 (0x00007fb0bb625000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fb0bb40f000)
/lib64/ld-linux-x86-64.so.2 (0x00007fb0d7740000)
libnvidia-fatbinaryloader.so.375.88 => /usr/lib/nvidia-375/libnvidia-fatbinaryloader.so.375.88 (0x00007fb0bb1c3000)

```

GCP P100 (has issue @ batch size 100)

libtensorflow_framework.so
```
linux-vdso.so.1 =>  (0x00007ffc665c5000)
libcublas.so.9.0 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcublas.so.9.0 (0x00007f25faa9f000)
libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007f25f9c22000)
libcudnn.so.7 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7 (0x00007f25e8786000)
libcufft.so.9.0 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcufft.so.9.0 (0x00007f25e06e5000)
libcurand.so.9.0 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcurand.so.9.0 (0x00007f25dc781000)
libcudart.so.9.0 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudart.so.9.0 (0x00007f25dc513000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f25dc30f000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f25dc006000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f25dbde8000)
libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f25dba66000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f25db850000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f25db485000)
/lib64/ld-linux-x86-64.so.2 (0x000056411f363000)
librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f25db27d000)
libnvidia-fatbinaryloader.so.384.90 => /usr/lib/nvidia-384/libnvidia-fatbinaryloader.so.384.90 (0x00007f25db02b000)
```

contrib/framework/python/ops/_variable_ops.so:
```
linux-vdso.so.1 =>  (0x00007ffe8276a000)
libtensorflow_framework.so => /opt/conda/envs/pytorch-py35/lib/python3.5/site-packages/tensorflow/contrib/f
ramework/python/ops/../../../../libtensorflow_framework.so (0x00007fd320a1a000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fd3206fd000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fd3204f9000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fd3202db000)
librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fd3200d3000)
libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fd31fd51000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd31f986000)
libcublas.so.9.0 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcublas.so.9.0 (0x00007fd31c54f000)
libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007fd31b6d1000)
libcudnn.so.7 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7 (0x00007fd30a236000)
libcufft.so.9.0 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcufft.so.9.0 (0x00007fd302195000)
libcurand.so.9.0 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcurand.so.9.0 (0x00007fd2fe230000)
libcudart.so.9.0 => /usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudart.so.9.0 (0x00007fd2fdfc3000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fd2fddad000)
/lib64/ld-linux-x86-64.so.2 (0x00005562f5375000)
libnvidia-fatbinaryloader.so.384.90 => /usr/lib/nvidia-384/libnvidia-fatbinaryloader.so.384.90 (0x00007fd2fdb5a000)

```
AWS V100 (does not have issue)

libtensorflow_framework.so
```
linux-vdso.so.1 =>  (0x00007ffdbc38a000)
/usr/lib/libmpi_cxx.so (0x00007fe10e8d1000)
libcublas.so.8.0 => /usr/local/cuda-8.0/lib64/libcublas.so.8.0 (0x00007fe10be39000)
libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007fe10afbc000)
libcudnn.so.6 => /usr/local/cuda-8.0/lib64/libcudnn.so.6 (0x00007fe101a5a000)
libcufft.so.8.0 => /usr/local/cuda-8.0/lib64/libcufft.so.8.0 (0x00007fe0f8c0b000)
libcurand.so.8.0 => /usr/local/cuda-8.0/lib64/libcurand.so.8.0 (0x00007fe0f4c95000)
libcudart.so.8.0 => /usr/local/cuda-8.0/lib64/libcudart.so.8.0 (0x00007fe0f4a2f000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fe0f482b000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fe0f4522000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe0f4305000)
libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fe0f3f83000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fe0f3d6d000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe0f39a3000)
/lib64/ld-linux-x86-64.so.2 (0x00007fe10f93c000)
libmpi.so.12 => /usr/lib/libmpi.so.12 (0x00007fe0f36cd000)
librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fe0f34c5000)
libnvidia-fatbinaryloader.so.384.81 => /usr/lib/x86_64-linux-gnu/libnvidia-fatbinaryloader.so.384.81 (0x00007fe0f3273000)
libibverbs.so.1 => /usr/lib/libibverbs.so.1 (0x00007fe0f3064000)
libopen-rte.so.12 => /usr/lib/libopen-rte.so.12 (0x00007fe0f2dea000)
libopen-pal.so.13 => /usr/lib/libopen-pal.so.13 (0x00007fe0f2b4d000)
libhwloc.so.5 => /usr/lib/x86_64-linux-gnu/libhwloc.so.5 (0x00007fe0f2913000)
libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007fe0f2710000)
libnuma.so.1 => /usr/lib/x86_64-linux-gnu/libnuma.so.1 (0x00007fe0f2505000)
libltdl.so.7 => /usr/lib/x86_64-linux-gnu/libltdl.so.7 (0x00007fe0f22fb000)

```
contrib/framework/python/ops/_variable_ops.so:
```
linux-vdso.so.1 =>  (0x00007ffdad35e000)
/usr/lib/libmpi_cxx.so (0x00007face59da000)
libtensorflow_framework.so => /home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/contrib/framework/python/ops/../../../../libtensorflow_framework.so (0x00007face4b8a000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007face4881000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007face467d000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007face4460000)
librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007face4258000)
libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007face3ed6000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007face3b0c000)
libmpi.so.12 => /usr/lib/libmpi.so.12 (0x00007face3836000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007face3620000)
libcublas.so.8.0 => /usr/local/cuda-8.0/lib64/libcublas.so.8.0 (0x00007face0b88000)
libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007facdfd0b000)
libcudnn.so.6 => /usr/local/cuda-8.0/lib64/libcudnn.so.6 (0x00007facd67a9000)
libcufft.so.8.0 => /usr/local/cuda-8.0/lib64/libcufft.so.8.0 (0x00007faccd95a000)
libcurand.so.8.0 => /usr/local/cuda-8.0/lib64/libcurand.so.8.0 (0x00007facc99e4000)
libcudart.so.8.0 => /usr/local/cuda-8.0/lib64/libcudart.so.8.0 (0x00007facc977e000)
/lib64/ld-linux-x86-64.so.2 (0x00007face5e87000)
libibverbs.so.1 => /usr/lib/libibverbs.so.1 (0x00007facc956f000)
libopen-rte.so.12 => /usr/lib/libopen-rte.so.12 (0x00007facc92f5000)
libopen-pal.so.13 => /usr/lib/libopen-pal.so.13 (0x00007facc9058000)
libnvidia-fatbinaryloader.so.384.81 => /usr/lib/x86_64-linux-gnu/libnvidia-fatbinaryloader.so.384.81 (0x00007facc8e06000)
libhwloc.so.5 => /usr/lib/x86_64-linux-gnu/libhwloc.so.5 (0x00007facc8bcc000)
libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007facc89c9000)
libnuma.so.1 => /usr/lib/x86_64-linux-gnu/libnuma.so.1 (0x00007facc87be000)
libltdl.so.7 => /usr/lib/x86_64-linux-gnu/libltdl.so.7 (0x00007facc85b4000)
```

**Comment by yzhwang** [comment]
@rwightman I did reproduce this bug using tf 1.4 built from source with python 3.5 on P100 + CUDA9 + cudnn7. The result is:
2017-11-20 10:58:07.427593: I tensorflow/core/kernels/logging_ops.cc:79] eval/Recall_5[0.76182]
2017-11-20 10:58:07.427593: I tensorflow/core/kernels/logging_ops.cc:79] eval/Accuracy[0.65166]
I will start looking into the problem.

**Comment by yzhwang** [comment]
@rwightman I spent some time with this and here are some updates:
All experiments were done using TensorFlow OSS version built from source.
I tried four settings with python 2.7/3.5 and CUDA8+cudnn6/CUDA9+cudnn7. Two settings with CUDA9+cudnn7 constantly reproduce the problem for me.
I can also confirm that transpose op and fused_batch_norm has no effects to the problem (switched to different kernel implementations and has no effect to final accuracy).
Using NCHW data format will have more cudnn conv2d calls in the model. For NHWC, there is only one cudnn conv2d call:
input: 100, 3, 331, 331
filter: 3, 96, 3, 3
stride: 2, 2
output: 100, 96, 165, 165

For NCHW, there are several other cudnn conv2d calls.
The difference doesn't affect the results when using CUDA8+cudnn6 though, so it could come from other places I haven't looked into such as separable depthwise conv.

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
